src/mbgl/style/layer.cpp

```diff
@@ -91,6 +91,59 @@ void Layer::setMaxZoom(float maxZoom) {
     observer->onLayerChanged(*this);
 }
 
+Value Layer::serialize() const {
+    mapbox::base::ValueObject result;
+    result.emplace("id", getID());
+
+    auto source = getSourceID();
+    if (!source.empty()) {
+        result.emplace("source", std::move(source));
+    }
+
+    if (!std::string(Layer::getTypeInfo()->type).empty()) {
```

Comment:
is it possible to have it empty?

Reply:
exceptional case for custom layer, type would be serialized as it is.